component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java

```diff
@@ -199,4 +209,117 @@ private List<String> populateMandatoryClaims() {
+     * @return Registered signing algorithm for the application.
+     * @throws OAuthClientAuthnException
+     */
+    private String getRegisteredSigningAlgorithm(String clientId) throws OAuthClientAuthnException {
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (!(isNotEmpty(serviceProvider.getCertificateContent()))) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (Constants.TOKEN_ENDPOINT_AUTH_SIGNING_ALG.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Token signing algorithm not registered",
+                    OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+        return Constants.NOT_APPLICABLE;
+    }
+
+    /**
+     * Obtain the request signing algorithm from the signed JWT.
+     *
+     * @param signedObject   Signed JWT in the request.
+     * @return Request signing algorithm.
+     * @throws OAuthClientAuthnException
+     */
+    private String getRequestSigningAlgorithm(String signedObject) throws OAuthClientAuthnException {
+        //retrieve algorithm from the signed JWT
+        try {
+            SignedJWT signedJWT = SignedJWT.parse(signedObject);
+            return signedJWT.getHeader().getAlgorithm().getName();
+        } catch (ParseException e) {
+            throw new OAuthClientAuthnException("Error occurred while parsing the signed assertion",
+                    OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+    }
+
+    /**
+     * Check whether the application should support FAPI.
+     *
+     * @param clientId       Client ID of the application.
+     * @return Whether the application should support FAPI.
+     * @throws OAuthClientAuthnException
+     */
+    private String isFapiApplication(String clientId) throws OAuthClientAuthnException {
+
+        String isFapiApp = "IsFAPIApp";
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (isEmpty(serviceProvider.getCertificateContent())) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (isFapiApp.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Token signing algorithm not registered",
```

Comment:
e is missing in throw

Reply:
IF we r skipping e clause for a reason,shall we least log it?

Reply:
Fixed in c18c85cc171d19c53412ac6eaad0e1e4b738d909